src/base/handlers.py

```diff
@@ -217,15 +218,43 @@ def dispatch(self):
     self._SetCommonResponseHeaders()
     super(BaseHandler, self).dispatch()
 
+
+  @classmethod
+  def get_jinja2_config(cls):
+    """
+      Builds Jinja2 config based on constants.
+
+      Note: this is used in the factory below, but an alternative way of setting
+      up Jinja2 would be to use the WSGIApplication config to set this and not
+      use the factory below. This has the advantage of having different settings
+      for different applications and not set here at the handler level.
+    """
+    extensions = ['jinja2.ext.with_']
+    return {
+      'environment_args': {
+        'autoescape': True,
+        'extensions': extensions,
+        'auto_reload': constants.DEBUG,
+      },
+      'template_path': constants.TEMPLATE_DIR
+    }
+
+  @staticmethod
+  def j2_factory(app):
+    """
+      The factory function passed to get_jinja2
+      Args:
+        app: the WSGIApplication
+    """
+    return jinja2.Jinja2(app, BaseHandler.get_jinja2_config())
+
   @webapp2.cached_property
   def jinja2(self):
-    extensions = ['jinja2.ext.with_']
-    env = jinja2.Environment(autoescape=True,
-                             auto_reload=constants.DEBUG,
-                             loader=jinja2.FileSystemLoader(
-                                 constants.TEMPLATE_DIR),
-                             extensions=extensions)
-    return env
+    """
+      Get the cached Jinja2 instance from the app registry, if none exists
+      the factory function is used to create one
```

Comment:
One last nit since I'm asking for another squash :-)  Can you add a '.' at the end of this sentence, and then squash everything (incl. contributor addition) prior to merge?